### PR TITLE
Behat tests for the incident threshold

### PR DESCRIPTION
**Pull request by solker**
#83

@amitaibu I rewrite tests, please check again.

### CODE REVIEW

File: behat/features/bootstrap/FeatureContext.php [hunk 1]
```diff
@@ -211,4 +231,160 @@ public function nodeOfTypeShouldBeDeleted($title, $type) {
...
+    if (!$ci_build_id) {
+      $params = [
+        '@title' => $ci_build
+      ];
+      throw new \Exception(format_string('Failed to get ID of CI build @title', $params));
+    }
+
+    // Get incident.
+    $query = new EntityFieldQuery();
+    $entity = $query->entityCondition('entity_type', 'node')
```
Comment (solker) [inline behat/features/bootstrap/FeatureContext.php]: I have problem with this query. In `devel/php` it is work but here I receive empty result.

File: behat/features/threshold.feature [hunk 2]
```diff
@@ -0,0 +1,36 @@
+Feature: Check threshold functionally
+  In order to be able to create CI build items with different states.
```
Comment (amitaibu) [inline behat/features/threshold.feature]: remove dots

File: behat/features/threshold.feature [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+  As a privileges user
+  I need to see the right status of CI build and new incidents.
+
+  @api
+  Scenario: Create two successfully CI build items and check status of CI build, it's should be 'Ok'.
+    Given I login with user "William"
+    When  I create repository and CI build "William/app1"
+    And   I set status "Done" for CI build "William/app1" items "2" times
```
Comment (amitaibu) [inline behat/features/threshold.feature]: `And "2" CI build items for CI build "William/app1" are set to status "Done"`

File: behat/features/threshold.feature [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+    Then  I should see status "Ok" for CI build "William/app1"
+
+  @api
+  Scenario: Create one failed CI build items and check status of CI build, it's should be 'Unconfirmed error'.
+    Given I login with user "William"
+    When  I create repository and CI build "William/app2"
+    And   I set status "Error" for CI build "William/app2" items "1" time
+    Then  I should see status "unconfirmed_error" for CI build "William/app2"
```
Comment (amitaibu) [inline behat/features/threshold.feature]: unconfirmed_error => `Unconfirmed error`

File: behat/features/threshold.feature [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+
+  @api @new
```
Comment (amitaibu) [inline behat/features/threshold.feature]: remove `@new`

File: behat/features/threshold.feature [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+  Scenario: Create two failed CI build items and check status of CI build, it's should be 'Error'.
+    Given I login with user "William"
+    When  I create repository and CI build "William/app3"
+    And   I set status "Error" for CI build "William/app3" items "2" times
+    Then  I should see status "error" for CI build "William/app3"
+
+  @api
```
Comment (amitaibu) [inline behat/features/threshold.feature]: split scenarios

File: behat/features/bootstrap/FeatureContext.php [hunk 7]
```diff
@@ -136,7 +136,27 @@ public function iCreateNodeOfType($title, $type) {
     $entity = entity_create('node', $values);
     $wrapper = entity_metadata_wrapper('node', $entity);
     $wrapper->title->set($title);
-    $wrapper->field_github_id->set(123456);
+    if ($type == 'repository') {
+      $wrapper->field_github_id->set(123456);
+    }
+    elseif ($type == 'ci_build') {
+      if (!$repository) {
+        $params = [
```
Comment (amitaibu) [inline behat/features/bootstrap/FeatureContext.php]: `array()`

File: behat/features/bootstrap/FeatureContext.php [hunk 8]
```diff
@@ -136,7 +136,27 @@ public function iCreateNodeOfType($title, $type) {
...
+      }
+
+      $query = new EntityFieldQuery();
+      $entity = $query->entityCondition('entity_type', 'node')
+        ->propertyCondition('title', $repository)
+        ->propertyCondition('type', 'repository')
+        ->range(0,1)
+        ->execute();
+
+      $repository_id = array_keys($entity['node'])[0];
```
Comment (amitaibu) [inline behat/features/bootstrap/FeatureContext.php]: `array_keys($entity['node'])[0]` => `key($result['node'])`

File: behat/features/bootstrap/FeatureContext.php [hunk 9]
```diff
@@ -136,7 +136,27 @@ public function iCreateNodeOfType($title, $type) {
...
+    elseif ($type == 'ci_build') {
+      if (!$repository) {
+        $params = [
+          '@title' => $title
+        ];
+        throw new \Exception(format_string('Failed to create a new CI build @title because repository is undefined.', $params));
+      }
+
+      $query = new EntityFieldQuery();
+      $entity = $query->entityCondition('entity_type', 'node')
```
Comment (amitaibu) [inline behat/features/bootstrap/FeatureContext.php]: $entity => $result

File: behat/features/bootstrap/FeatureContext.php [hunk 10]
```diff
@@ -136,7 +136,27 @@ public function iCreateNodeOfType($title, $type) {
...
+        ->propertyCondition('title', $repository)
+        ->propertyCondition('type', 'repository')
```
Comment (amitaibu) [inline behat/features/bootstrap/FeatureContext.php]: `->entityCondition('entity_type', 'node')` => `->entityCondition('bundle', 'repository')`

File: behat/features/bootstrap/FeatureContext.php [hunk 11]
```diff
@@ -211,4 +231,165 @@ public function nodeOfTypeShouldBeDeleted($title, $type) {
     );
     throw new \Exception(format_string("Node @title of @type was found.", $params));
   }
+
+  /**
+   * @When I create repository and CI build :title
+   */
+  public function iCreateRepositoryAndCiBuild($title) {
+    // First, create a new repository.
```
Comment (amitaibu) [inline behat/features/bootstrap/FeatureContext.php]: remove First, and Second.

File: behat/features/bootstrap/FeatureContext.php [hunk 12]
```diff
@@ -211,4 +231,165 @@ public function nodeOfTypeShouldBeDeleted($title, $type) {
...
+    $this->iCreateNodeOfType($title, 'repository');
+    // Second, create a new CI build.
+    $this->iCreateNodeOfType($title, 'ci_build', $title);
+  }
+
+  /**
+   * @When /^I set status "([^"].+)" for CI build "([^"].+)" items "([0-9]+)" times?$/
+   */
+  public function iSetStatusForCiBuildItemsTimes($status, $ci_build, $times) {
+    if ($times <= 0) return;
```
Comment (amitaibu) [inline behat/features/bootstrap/FeatureContext.php]:
``` php
if ($times <= 0) {
  return;
}
```

File: behat/features/bootstrap/FeatureContext.php [hunk 13]
```diff
@@ -211,4 +231,165 @@ public function nodeOfTypeShouldBeDeleted($title, $type) {
...
+
+    // Get ID of CI build.
+    $query = new EntityFieldQuery();
+    $entity = $query->entityCondition('entity_type', 'node')
```
Comment (amitaibu) [inline behat/features/bootstrap/FeatureContext.php]: $result

File: behat/features/bootstrap/FeatureContext.php [hunk 14]
```diff
@@ -211,4 +231,165 @@ public function nodeOfTypeShouldBeDeleted($title, $type) {
...
+      ];
+      throw new \Exception(format_string('Failed to get ID of CI build @title', $params));
+    }
+
+    // Find last CI build item.
+    $query = new EntityFieldQuery();
+    $entity = $query->entityCondition('entity_type', 'message')
+      ->fieldCondition('field_ci_build_status', 'value', 'queue')
+      ->fieldCondition('field_ci_build', 'target_id', $ci_build_id)
+      ->range(0,1)
```
Comment (amitaibu) [inline behat/features/bootstrap/FeatureContext.php]: remove, to do a count check

File: behat/features/bootstrap/FeatureContext.php [hunk 15]
```diff
@@ -211,4 +231,165 @@ public function nodeOfTypeShouldBeDeleted($title, $type) {
...
+      ->execute();
+
+    // TODO: Check that count of CI build items only one.
+
+    $message = message_load(array_keys($entity['message'])[0]);
```
Comment (amitaibu) [inline behat/features/bootstrap/FeatureContext.php]:
``` php
$mid = key(...
$wrapper = entity_metadata_wrapper('message', $mid);
```

File: behat/features/bootstrap/FeatureContext.php (hunk 12 above)
Comment (amitaibu) [inline behat/features/bootstrap/FeatureContext.php]: add comment

File: behat/features/bootstrap/FeatureContext.php [hunk 16]
```diff
@@ -211,4 +231,165 @@ public function nodeOfTypeShouldBeDeleted($title, $type) {
...
+
+    // Get last incident for CI build.
+    $query = new EntityFieldQuery();
+    $entity = $query->entityCondition('entity_type', 'node')
+      ->propertyCondition('type', 'ci_incident')
+      ->fieldCondition('field_ci_build', 'target_id', $ci_build_id)
+      ->range(0, 1)
+      ->execute();
+
+    var_dump($ci_build_id);
```
Comment (amitaibu) [inline behat/features/bootstrap/FeatureContext.php]: remove

File: behat/features/bootstrap/FeatureContext.php [hunk 17]
```diff
@@ -211,4 +231,165 @@ public function nodeOfTypeShouldBeDeleted($title, $type) {
...
+    $ci_build_id = array_keys($entity['node'])[0];
+    if (!$ci_build_id) {
+      $params = [
+        '@title' => $ci_build
+      ];
+      throw new \Exception(format_string('Failed to get ID of CI build @title', $params));
+    }
+
+    // Get last incident for CI build.
+    $query = new EntityFieldQuery();
```
Comment (amitaibu) [inline behat/features/bootstrap/FeatureContext.php]: use `shoov_ci_incident_get_latest_error_incident()`

File: behat/features/bootstrap/FeatureContext.php [hunk 18]
```diff
@@ -211,4 +231,165 @@ public function nodeOfTypeShouldBeDeleted($title, $type) {
...
+    if ($failed_count != $count) {
+      $params = [
+        '@title' => $ci_build,
+        '@failed_count' => $failed_count,
+        '@count' => $count
+      ];
+      throw new \Exception(format_string('CI build @title have failed count equal to @failed_count instead of @count', $params));
+    }
+  }
+
```
Comment (amitaibu) [inline behat/features/bootstrap/FeatureContext.php:332]:
``` php
  protected function getNodeByTitleAndBundle($title, $bundle) {
    $query = new \entityFieldQuery();
    $result = $query
      ->entityCondition('entity_type', 'node')
      ->entityCondition('bundle', $bundle)
      ->propertyCondition('title', $title)
      ->propertyCondition('status', NODE_PUBLISHED)
      ->range(0, 1)
      ->execute();

    // ...
  }
```

File: behat/features/bootstrap/FeatureContext.php [hunk 19]
```diff
@@ -211,4 +231,165 @@ public function nodeOfTypeShouldBeDeleted($title, $type) {
...
+    // Predefine properties for exception.
+    $params = [
+      '@title' => $ci_build,
+      '@status' => $status
+    ];
+    $error_message = format_string('The incident for CI build @title doesn\'t contain "@status" item', $params);
+
+    if ($status == 'error') {
+      $failing_build = $wrapper->field_failing_build->value();
```
Comment (amitaibu) [inline behat/features/bootstrap/FeatureContext.php]: if (!$failing_build = $wrapper->field_failing_build->value()) {
}

File: behat/features/bootstrap/FeatureContext.php (hunk 10 above)
Comment (solker) [inline behat/features/bootstrap/FeatureContext.php]: @amitaibu 
`propertyCondition` doesn't work without `entity_type`.

``` php
  protected function propertyQuery() {
    if (empty($this->entityConditions['entity_type'])) {
      throw new EntityFieldQueryException(t('For this query an entity type must be specified.'));
    }
    ...
```

I tried get title via `fieldCondition` but got fail.

Comment (amitaibu) [inline behat/features/bootstrap/FeatureContext.php]: Sorry, I wrote it wrong, this is what I meant:

``` php
$result = $query
  ->entityCondition('entity_type', 'node')
  ->entityCondition('bundle', 'repository')
  // ...
  ->execute();
```

**Comment by amitaibu** [comment]
Merged. @solker you can look at the fixes I've made.

**Pull request closed without merging**